src/number/complete.rs

```diff
@@ -1355,31 +1355,32 @@ where
 ///   hex_u32(s)
 /// };
 ///
-/// assert_eq!(parser(&b"01AE"[..]), Ok((&b""[..], 0x01AE)));
-/// assert_eq!(parser(&b"abc"[..]), Ok((&b""[..], 0x0ABC)));
-/// assert_eq!(parser(&b"ggg"[..]), Err(Err::Error((&b"ggg"[..], ErrorKind::IsA))));
+/// assert_eq!(parser("01AE"), Ok(("", 0x01AE)));
+/// assert_eq!(parser("abc"), Ok(("", 0x0ABC)));
+/// assert_eq!(parser("ggg"), Err(Err::Error(("ggg", ErrorKind::HexDigit))));
+/// assert_eq!(parser(""), Err(Err::Error(("", ErrorKind::HexDigit))));
 /// ```
 #[inline]
-pub fn hex_u32<'a, E: ParseError<&'a [u8]>>(input: &'a [u8]) -> IResult<&'a [u8], u32, E> {
-  let (i, o) = crate::bytes::complete::is_a(&b"0123456789abcdefABCDEF"[..])(input)?;
+pub fn hex_u32<'a, I, E: ParseError<I>>(input: I) -> IResult<I, u32, E>
+where
+  I: InputIter,
+  I: Slice<RangeFrom<usize>> + Slice<RangeTo<usize>>,
+  <I as InputIter>::Item: AsChar,
+{
   // Do not parse more than 8 characters for a u32
-  let (parsed, remaining) = if o.len() <= 8 {
-    (o, i)
+  let mut chars = input
+    .iter_elements()
+    .take(8)
+    // Can replace map-take_while-flatten with `.map_while` once stabilized
+    .map(|c| c.as_char().to_digit(16))
+    .take_while(|digit| digit.is_some())
+    .flatten();
+  if let Some(first) = chars.next() {
+    let (num, len) = chars.fold((first, 1), |(acc, len), item| (16 * acc + item, len + 1));
```

Comment:
I don't love this for readability. It does work.
Might be better to just call collect() on `chars` as before - to collect the up-to-8 chars for the hex number.